core/server/web/shared/middlewares/image/handle-image-sizes.js

```diff
@@ -0,0 +1,59 @@
+const path = require('path');
+const sharp = require('sharp');
+const storage = require('../../../../adapters/storage');
+const activeTheme = require('../../../../services/themes/active');
+
+const SIZE_PATH_REGEX = /^\/size\/([^/]+)\//;
+
+module.exports = function (req, res, next) {
+    if (!SIZE_PATH_REGEX.test(req.url)) {
+        return next();
+    }
+
+    const imageSizes = activeTheme.get().config('image_sizes');
+    const imageDimensions = Object.keys(imageSizes).reduce((dimensions, size) => {
+        const {width, height} = imageSizes[size];
+        const dimension = (width ? 'w' + width : '') + (height ? 'h' + height : '');
+        return Object.assign({
+            [dimension]: imageSizes[size]
+        }, dimensions);
+    }, {});
+
+    const [sizeImageDir, requestedDimension] = req.url.match(SIZE_PATH_REGEX);
+
+    // CASE: unknown size or missing size config
+    const imageDimensionConfig = imageDimensions[requestedDimension];
+    if (!imageDimensionConfig || (!imageDimensionConfig.width && !imageDimensionConfig.height)) {
+        const url = req.originalUrl.replace(`/size/${requestedDimension}`, '');
+        return res.redirect(url);
+    }
+
+    const storageInstance = storage.getStorage();
+    // CASE: unsupported storage adapter but theme is using custom image_sizes
+    if (typeof storageInstance.saveRaw !== 'function') {
+        const url = req.originalUrl.replace(`/size/${requestedDimension}`, '');
+        return res.redirect(url);
+    }
+
+    storageInstance.exists(req.url).then((exists) => {
+        if (exists) {
+            return;
+        }
+
+        const originalImagePath = path.relative(sizeImageDir, req.url);
+
+        return storageInstance.read({path: originalImagePath})
+            .then((originalImageBuffer) => {
```

Comment:
I think it would help the readability if this part was written on the main `.then` chain = less nesting+better readability of steps that are taken. Something like this:
```
    storageInstance.exists(req.url)
        .then((exists) => {
            if (exists) {
                return;
            }

            const originalImagePath = path.relative(sizeImageDir, req.url);

            return storageInstance.read({path: originalImagePath});
        })
        .then((originalImageBuffer) => {
            return sharp(originalImageBuffer)
                .resize(imageDimensionConfig.width, imageDimensionConfig.height, {
                    withoutEnlargement: true
                })
                .toBuffer();
        })
        .then((resizedImageBuffer) => {
            return storageInstance.saveRaw(resizedImageBuffer, req.url);
        })
        .then(() => {
            next();
        })
        .catch(next);
```

Reply:
The trouble with that is the early exit for the `if (exists)` condition - by pulling the inner promise out we completely change the control flow, and then would have to ass some extra custom handling for skipping